Add instructions for working with merge requests

linduxed opened the pull request:
Covers creating, working in and closing merge requests in Gitlab.

linduxed committed:
Add instructions for working with merge requests

Covers creating, working in and closing merge requests in Gitlab.

calleluks commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,104 @@
+   **Title and Description guidelines** section.  
+   Set no label, this will be done by someone else.  
+   Set no assignee, unless you've been asked to.
+9. Press the "Submit merge request" button.
```
Should developers be encouraged to review their changes themselves before submitting the MR?

linduxed replied:
Personally, I would want to see that included. However, I'm not sure if including this kind of directive would lead to apprehensiveness against submitting the MRs. I'd like to hear what others think first.

jeltz replied:
Perhaps mention that people _can_ review the MR here before submitting it.

linduxed replied:
This is a good way of formulating it! I'll use it.

calleerlandsson commented:
Pedantry: `s/Gitlab/GitLab/g`

jeltz commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,104 @@
+become easier to if you surround it with backticks.
+
+## Closing a merge requests
+
```
Remove this extra empty line..

jeltz commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,104 @@
+
+When the reviewer considers the code ready to merge, he'll say so, and a label
+will be applied that indicates that the branch is ready to merge and deploy.
+Don't worry about the label, someone else will handle applying it.
```
I think we should also inform reviewers about what this label should be.

linduxed replied:
I chose this wording because to my knowledge the name of the label has already changed two times, maybe even three. Until it's set in stone (or at least a bit less volatile), I think we can rely on word of mouth, since the only people applying the labels are the select few that have admin rights.

jeltz commented:
Should we suggest that you add a link to the MR in Redmine. I always do this.

linduxed commented:
On line 47 there should be an instruction saying that just that.

linduxed committed:
Add section about finding merge requests

linduxed commented:
Ready for another round of review.

linduxed commented:
One thing that @jsageryd pointed out that I'd like to hear some opinions on is the creation of fixup commits that you only squash at the very end of the life of a MR.

If we disregard the fact that lately GitLab seems to act in such a way that it would seem that it's fine if we rebase and squash things any way we like (it seems comments and most other information stays around as it should), wouldn't it be better to include this kind of instructions over in the [workflow part of the git style guide](../tree/master/style/git#workflow)?

edruid commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,116 @@
+become easier to if you surround it with backticks.
+
+### Closing a merge requests
+When the reviewer considers the code ready to merge, he'll say so, and a label
```
It would be nice to have a document detailing what to do as reviewer linked here detailing what labels should be applied and such.

edruid replied:
Actually I wouldn't mind stating the actual label here if it is only one label just so you know what to look for when the review is done.

linduxed replied:
As of now, only certain people (those with admin rights) can apply the labels, so this will most likely not be relevant to the people reading this. It's a bit along the lines of "if you're allowed to do it, you'll know the labels".

With that said, the label we're currently applying to the MRs that have been approved for merging get the label `ready-for-merge`. I'll add this to the document.

linduxed replied:
This should now be addressed.

jsageryd replied:
s/he'll say so/he will add a comment to say so/ for clarity?

jsageryd replied:
And I agree with Eric's comment earlier, it would be nice to add more info on what to do as a reviewer, but perhaps we can find what works best after some time. But, how about right now introducing the voting system (reviewer writes +1 in comment) for an easy way to spot whether or not something is ready? At least one +1 needed to get something merged, for example.

edruid commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,116 @@
+### Closing a merge requests
+When the reviewer considers the code ready to merge, he'll say so, and a label
+will be applied that indicates that the branch is ready to merge and deploy.
+Don't worry about the label, someone else will handle applying it.
```
I'd like to at least have the role of the person applying the label so I know which tree to bark up when no labels have been added.

linduxed replied:
Good point, I'll change this.

linduxed committed:
Add label name to merge request process document

Additionally, tweak the surrounding sentences for easier reading.

linduxed committed:
Add info about who can apply labels in GitLab

linduxed commented:
Ready for another round of review.

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+
+These are the steps necessary:
+
+1. Log into GitLab.
```
s/into/in to/

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+#### Title and Description guidelines
+The title should always look something like this:
+
+    #12345: Adjust constant in warpcore.yml to prevent overheating
```
s/#12345:/#12345/ ? I commented on this earlier but I don't know what happened to my comment.

jsageryd replied:
Also, in accordance with commit style guide, avoid mentioning what files are changed but rather mention e.g. what part of the project is changed, perhaps just s/.yml// would do fine.

linduxed replied:
I'll repeat my previous response then: I think that it looks both more readable and more aesthetically pleasing to keep the colon there. I don't consider it to be confusing or prone to error. Even if people do the mistake of dropping it, it's not a big deal (and if the majority of us try to include the colon, people will adjust automatically, noticing a pattern).

linduxed replied:
As for the file name, I'll remove it.

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+
+* What the MR changes, being a bit more descriptive than the title.
+* If there's something you're unsure of how it should be done, point this out so
+  the reviewer can help you out.
```
s/so the reviewer/so that the reviewer/

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+* What the MR changes, being a bit more descriptive than the title.
+* If there's something you're unsure of how it should be done, point this out so
+  the reviewer can help you out.
+* If necessary, include information that would help the reviewers understand
```
s/reviewers/reviewer/ for consistency

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+The process of working within MRs doesn't differ much from how review has been
+handled over in Redmine, with one major difference:
+
+If you make an adjustment (eg. to address a comment that was left by
```
s/eg./e.g./

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+in the MR.
+
+Additionally, commenting on specific lines in the code is very easy, so feel
+free to use the feature whenever it's applicable. To comment on a line, press
```
s/whenever it's applicable/whenever applicable/

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+
+Additionally, commenting on specific lines in the code is very easy, so feel
+free to use the feature whenever it's applicable. To comment on a line, press
+the "Changes" tab near the top of the MR, then press a line number in one of the
```
I don't think we should encourage commenting on entire change-sets, reviewing on a per-commit basis has benefits:
- Implementor takes more care to make proper logical commits
- Implementor understands more easily what is being commented on and therefore what commit to reference in his fixup commit
- Reviewer can review changes in smaller chunks

linduxed replied:
First of all, I'd like to address the first point you raise:
Your assertion that the implementor would take care to write proper logical commits for the benefit of the reviewer strikes me as odd. How we should structure our commits is already something that is detailed by the Git style guide and it's done for more reasons than just the review process.

As for the following comments I'll say the following:

Reviewing commit by commit works well when the commits almost completely stand on their own. To be fair, for our commits this is often the case, but not always. If there would be a single commit that isn't squashed into a previous one that patches the logic, or just changes the concept/idea of the code somewhat (all in the same branch that is being reviewed), then this is immediately more "reviewable" in the view that simply shows the sum of all changes.

While I understand the merits of reviewing commit after commit, I interpret "I don't think we should encourage" as you thinking usage of the "Changes" view is something negative and this I can't agree with.

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+Everywhere where you can leave a comment,
+[GitLab flavoured Markdown](http://doc.gitlab.com/ce/markdown/markdown.html) is
+used. While it's helpful if you glance through the entire document, if you only
+take away one thing from that it, it should be that you can turn any text into
```
s/from that it,/from it,/

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+called: `` ` ``). If you're talking about `SomeClass.with.some_method`, things
+become easier to if you surround it with backticks.
+
+### Closing a merge requests
```
s/requests/request/

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,118 @@
+Once it has been said in the MR that it's ready for merging, write the same in
+Redmine (if the reviewer hasn't done it already).
+
+You don't need to close the MR, it will be closed automatically when the branch
```
s/You don't need to close the MR/Do not close the MR/

jsageryd commented:
About the fixup commits that you wrote about above, @linduxed, I think they do belong in this guide because as far as our Git workflow goes, if we did not have reviews, we would never have to preserve fixup commits or even state rules for how to use them. And I think that it is important to include in this guide, otherwise people will force-push everything all the time, chaos and confusion arises.

linduxed commented:
This is a good point, I'll introduce the instructions regarding the fixup commits.

linduxed committed:
Fix typos and minor issues

Address comments raised during review, mostly typos and grammatical
errors.

linduxed committed:
Add info about how to work with git in MRs

Necessary since the aversion towards rebasing frequently differs from
our previous workflow.

linduxed committed:
Minor markdown style fix

linduxed commented:
Ready for another review.

jsageryd commented on process/merge_requests/README.md:
```diff
@@ -0,0 +1,141 @@
+What does this mean?
+
+Let's say you've submitted your MR and the reviewer wants you to change
+something in a previous commit. If you previously had the habit of creating
```
👌

jsageryd commented:
I think this looks good to merge.
